File: src/main/kotlin/day3/Part2.java
package day3;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.stream.Stream;

public class Part2 {
    public static void main(String[] args) throws IOException {
        List<String> strings = Files.readAllLines(Paths.get("src/main/resources/day3.txt"));

        Long count = Stream.of(
                slope(strings, 1, 1),
                slope(strings, 3, 1),
                slope(strings, 5, 1),
                slope(strings, 7, 1),
                slope(strings, 1, 2)
        ).reduce((aLong, aLong2) -> aLong * aLong2)
                .orElse(0L);
        System.out.println(count);
    }

    private static long slope(List<String> strings, int right, int down) {
        int trees = 0;
        for (int i = 0, stepsRight = 0; i < strings.size(); i = i + down, stepsRight += right) {
            int x = stepsRight % strings.get(i).length();
            if (strings.get(i).charAt(x) == '#') {
                trees++;
            }
        }
        return trees;
    }
}

File: src/main/kotlin/day3/Part2.kt
package day3

import java.nio.file.Files
import java.nio.file.Paths

fun main() {
    val lines = Files.readAllLines(Paths.get("src/main/resources/day3.txt"))

    val count = arrayOf(
        slope(lines, 1, 1),
        slope(lines, 3, 1),
        slope(lines, 5, 1),
        slope(lines, 7, 1),
        slope(lines, 1, 2),
    ).map { println(it); it }.reduce { acc, i -> acc * i }

    println(count)
}

private fun slope(lines: MutableList<String>, right: Int, down: Int, ): Int {
    var stepRight = 0;
    return lines
        .filterIndexed { index, s -> index % down == 0 }
        .filter { line ->
            val posX = stepRight % line.length;
            stepRight += right;
            line[posX] == '#'
        }.count()
}